### PR TITLE
Have multiple realms show as folders within one main folder

### PR DESCRIPTION
**Pull request by FadhlanR**
In this PR, each realm that the user selects to open through the 'Attach to Boxel Workspaces' command will be opened under a root folder named 'Boxel Workspaces'.  To ensure uniqueness and clarity, the directory name for each realm will be a combination of the realm name and the realm path. Since the realm path is unique, the resulting directory name will also be unique.

Example:

Realm Name: Draft Workspace
Realm Path: /john/draft-workspace
Directory Name: Draft Workspace (john-draft-workspace)

https://github.com/user-attachments/assets/ccc11e29-83a7-46d4-90e1-cab7e2761466

### CODE REVIEW

File: packages/vscode-boxel-tools/src/file-system-provider.ts [hunk 1]
```diff
@@ -57,12 +52,47 @@ export class Directory implements vscode.FileStat {
...
 
+  private context: vscode.ExtensionContext;
+  private selectedRealms: Set<string> = new Set();
+  private realmDirNamePromises: Map<string, Promise<string>> = new Map();
+  private realmDirNameToUrl: Map<string, string> = new Map();
+
+  constructor(context: vscode.ExtensionContext) {
+    this.context = context;
+
+    let selectedRealms = this.context.globalState.get<string>(selectedRealmKey);
```
Comment (IanCal) [inline packages/vscode-boxel-tools/src/file-system-provider.ts:70]: Do we need to store this state? Does something happen if it's only in this.selectedRealms?

**Review by IanCal — COMMENTED** [review]
Tested in cursor, works nicely and supports the composer which was the key thing we wanted.

I like explicitly adding new workspaces rather than automatically loading all.

Just an open question around the use of the storage, and if we need it.

File: packages/vscode-boxel-tools/src/file-system-provider.ts (hunk 1 above)
Comment (FadhlanR) [reply packages/vscode-boxel-tools/src/file-system-provider.ts:70]: I've tried not to use this state, but it seems that the activate function can be called multiple times in a workspace. This causes realmFs to be re-initialized, resulting in the in-memory state, such as selectedRealms, being reset. I don't know if there's better approach.

**Comment by backspace** [comment]
I’m sure it’s just some weird local development chaos but I’m unable to get this working:

extension-development-host-workspace code-workspace 2024-10-22 17-18-17

I was logging in with a user that doesn’t have access to `experiments`

I think there’s some caching or something that’s a deeper problem. But I’ll try again tomorrow.

File: packages/vscode-boxel-tools/src/file-system-provider.ts (hunk 1 above)
Comment (IanCal) [reply packages/vscode-boxel-tools/src/file-system-provider.ts:70]: Ah yes. Extensions are reactivated if the first folder in the first folder is changed in the workspace, so I think we typically get two activations.

We might be able to get away with a workspace configuration - we don't need it to be stored between windows right? That could also be related to the bug @backspace saw, but I'm not sure.

Comment (FadhlanR) [reply packages/vscode-boxel-tools/src/file-system-provider.ts:70]: I've added comments as we discussed [747b77f](https://github.com/cardstack/boxel/pull/1709/commits/747b77f4944b6195490c514c690e1b9c2883eeb6)

**Review by backspace — APPROVED** [review]
Who knows why it would be different now but it works for me locally!

**Pull request merged**